src/fn_call.rs

```diff
@@ -152,8 +159,151 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
+                // }
+                //
+                // box_me_up is the first method in the vtable.
+                // First, we extract the vtable pointer from our fat pointer,
+                // and check its alignment
+
+                let vtable_ptr = payload_dyn.to_meta()?.expect("Expected fat pointer!").to_ptr()?; 
+                let data_ptr = payload_dyn.to_scalar_ptr()?;
+                this.memory().check_align(vtable_ptr.into(), this.tcx.data_layout.pointer_align.abi)?;
+
+                // Now, we dereference the vtable pointer.
+                let alloc = this.memory().get(vtable_ptr.alloc_id)?;
+
+                // Finally, we extract the pointer to 'box_me_up'.
+                // The vtable is layed out in memory like this:
+                //
+                //```
+                // <drop_ptr> (usize)
+                // <size> (usize)
+                // <align> (usize)
+                // <method_ptr_1> (usize)
+                // <method_ptr_2> (usize)
+                // ...
+                // <method_ptr_n> (usize)
+                //```
+                //
+                // Since box_me_up is the first method pointer
+                // in the vtable, we use an offset of 3 pointer sizes
+                // (skipping over <drop_ptr>, <size>, and <align>)
+
+                let box_me_up_ptr = alloc.read_ptr_sized(
+                    this,
+                    vtable_ptr.offset(this.pointer_size() * 3, this)?
+                )?.to_ptr()?;
+
+                // Get the actual function instance
+                let box_me_up_fn = this.memory().get_fn(box_me_up_ptr)?;
+                let box_me_up_mir = this.load_mir(box_me_up_fn.def)?;
+
+                // Extract the signature
+                // We know that there are no HRBTs here, so it's fine to use
+                // skip_binder
+                let fn_sig_temp = box_me_up_fn.ty(*this.tcx).fn_sig(*this.tcx);
+                let fn_sig = fn_sig_temp.skip_binder();
+
+                // This is the layout of '*mut (dyn Any + Send)', which
+                // is the return type of 'box_me_up'
+                let dyn_ptr_layout = this.layout_of(fn_sig.output())?;
+
+                // We allocate space to store the return value of box_me_up:
+                // '*mut (dyn Any + Send)', which is a fat 
+                
+                let temp_ptr = this.allocate(dyn_ptr_layout, MiriMemoryKind::UnwindHelper.into());
+                this.machine.box_me_up_tmp_ptr = Some(temp_ptr.clone());
+
+                this.push_stack_frame(
+                    box_me_up_fn,
+                    box_me_up_mir.span,
+                    box_me_up_mir,
+                    Some(temp_ptr.into()),
+                    // We know that this frame will always be unwound,
+                    // since we enable unwinding as soon as we encounter it
+                    // in Evaluator::stack_pop
+                    StackPopCleanup::Goto { ret: None, unwind }
```

Comment:
That seems weird. Unwinding a frame should happen when a function panicked, but this one did not panic, right? It should always return normally -- except it should return to our unwinding path here. If you just make this the "return" edge, but make it return into our unwind path, shouldn't that work?

Reply:
This is kind of weird. In `libpanic_unwind`, there's [additional code that runs](https://github.com/rust-lang/rust/blob/81eb15210a0c43820b146cd9ae6fd2062e1b2922/src/libpanic_unwind/lib.rs#L101) with the rest of `box_me_up`. However, in Miri, there's nothing else that we could possibly run - we need to immediately jump to the unwind block.

Since we want rustc's unwinding logic to start running as soon as we pop off this frame, we need to use `unwind`. I could special-case the start of unwinding in `rustc`, but I thought it would be simpler to have rustc be oblivious to the details of the `box_me_up` hack.

Reply:
Couldn't that extra code run in `stack_pop` similar to how you run extra stuff for `catch_panic`?

Reply:
That's exactly what I do. However, we need to set an extra flag so that `stack_pop` knows when it has encountered the target frame.

Reply:
Right but what does that have to do with setting `ret: None` and only setting the `unwind` edge.

This might have to do with the unwind flag, which we are still discussing in the rustc PR. So maybe let's first complete review of the rustc PR before continuing here.